### PR TITLE
Feature/all nullable parquet writer

### PR DESCRIPTION
**Pull request by felipemmelo**

### CODE REVIEW

File: ingestor/shared/src/test/scala/za/co/absa/hyperdrive/shared/utils/TestSparkTestUtils.scala [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  private val nestedTypeTrueNullability = new StructType()
+    .add(StructField("e", ArrayType(flatTypeTrueNullability, containsNull = true), nullable = true))
+
+  private val nestedTypeFalseNullability = new StructType()
+    .add(StructField("f", ArrayType(flatTypeFalseNullability, containsNull = false), nullable = false))
+
+  private val nestedTrueButOthersFalse = new StructType()
+    .add(StructField("g", ArrayType(flatTypeFalseNullability, containsNull = true), nullable = false))
+
+  describe("allFieldsContainNullability") {
```
Comment (kevinwallimann) [inline ingestor/shared/src/test/scala/za/co/absa/hyperdrive/shared/utils/TestSparkTestUtils.scala]: I think you can omit the test cases for `allFieldsContainNullability` and make this method `private`. The test cases for `allFieldsContainNullability` are fully covered by the test cases for `areAllFieldsNonNullable` and `areAllFieldsNullable` together.

File: ingestor/shared/src/test/scala/za/co/absa/hyperdrive/shared/utils/TestSparkUtils.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    val nestedType1 = new StructType()
+      .add(StructField("a", IntegerType, nullable = false))
+      .add(StructField("b", LongType, nullable = false))
+
+    val nestedType2 = new StructType()
+      .add(StructField("c", ArrayType(nestedType1, containsNull = false), nullable = false))
+
+    val schema = new StructType()
+      .add(StructField("d", StringType, nullable = false))
+      .add(StructField("e", ArrayType(nestedType2, containsNull = false), nullable = false))
```
Comment (kevinwallimann) [inline ingestor/shared/src/test/scala/za/co/absa/hyperdrive/shared/utils/TestSparkUtils.scala]: 👍 for named attributes

File: ingestor/shared/src/main/scala/za/co/absa/hyperdrive/shared/utils/SparkUtils.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.spark.sql.types.{ArrayType, StructField, StructType}
+
+object SparkUtils {
+
+  def setAllColumnsNullable(df: DataFrame): DataFrame = {
+    val newSchema = setSchemaNullability(df.schema, desiredNullability = true)
+    df.sparkSession.createDataFrame(df.rdd, newSchema)
+  }
+
+  def setAllColumnsNonNullable(df: DataFrame): DataFrame = {
```
Comment (kevinwallimann) [inline ingestor/shared/src/main/scala/za/co/absa/hyperdrive/shared/utils/SparkUtils.scala]: Why did you add this method, because it is unused?

File: ingestor/shared/src/test/scala/za/co/absa/hyperdrive/shared/utils/TestSparkTestUtils.scala (hunk 1 above)
Comment (felipemmelo) [reply ingestor/shared/src/test/scala/za/co/absa/hyperdrive/shared/utils/TestSparkTestUtils.scala]: True, will do it.

File: ingestor/shared/src/main/scala/za/co/absa/hyperdrive/shared/utils/SparkUtils.scala (hunk 3 above)
Comment (felipemmelo) [reply ingestor/shared/src/main/scala/za/co/absa/hyperdrive/shared/utils/SparkUtils.scala]: It was for completeness, i.e. nullable and non-nullable. On the other hand, it breaks YAGNI. I agree with you and will go for YAGNI for now.
